### PR TITLE
POC for segment replication.

### LINKED ISSUE #1544: Develop a proof-of-concept implementation for Segment Replication
**Is your feature request related to a problem? Please describe.**
This is a follow up task to https://github.com/opensearch-project/OpenSearch/issues/1123

**Describe the solution you'd like**
Develop a proof-of-concept (POC) of segment replication in a feature branch

**Describe alternatives you've considered**
N/A

**Additional context**
https://github.com/opensearch-project/OpenSearch/issues/1123

### PR DESCRIPTION
**Pull request by mch2**
Signed-off-by: Marc Handalian <handalm@amazon.com>

### Description
In this POC replicas are configured as read only.
After primary shard refresh, a checkpoint is published over the transport layer to replicas.
Once received, replicas fetch files in the checkpoint from the primary shard.
This initial commit ignores failover, retention leases, and shard allocation.

There are numerous locations where code is commented out with TODOs.  Gradle check and ITs are expected to fail.

Guide to reading this - 
1.  When a replica is created it now starts using `SegmentReplicationService` which is similar to PeerRecoveryTargetService but uses a `ReplicationTarget` class instead of a RecoveryTarget to drive operations on the replica shard.
2. `ReplicationTarget` will bootstrap the shard and request checkpoint information, missing/changed files, and the `SegmentInfos` bytes from the primary.  After files are received, it will update the current reader's SegmentInfos inside `OpenSearchReaderManager` and refresh.   This step is the same as Lucene's [SegmentInfosSearcherManager](https://github.com/apache/lucene/blob/48715fe898bf0a900e3e6b681882ecb1b6d8e1b1/lucene/replicator/src/java/org/apache/lucene/replicator/nrt/SegmentInfosSearcherManager.java#L50).
3. On the Primary shard, `SegmentReplicationPrimaryService` handles requests to fetch checkpoints & send files.  It delegates to `PrimaryShardReplicationHandler` to send files - similar to recovery's RecoverySourceHandler.  SegmentInfos is fetched directly from the current reader running on the primary.
4.   A Refresh listener `CheckpointRefreshListener` publishes a new checkpoint over the transport layer using `TransportCheckpointPublisher` when the primary shard refreshes. 
5.  Checkpoints are sent to replicas and received in `TransportPublishShardCheckpointAction`, when received it invokes `IndexShard.onNewCheckpoint` to kick off another replication with `SegmentReplicationService`.

### Issues Resolved
closes #1544 
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/indices/replication/copy/ReplicationCollection.java [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+    public ReplicationRef getReplicationSafe(long id, ShardId shardId) {
+        ReplicationRef replicationRef = getReplication(id);
+        if (replicationRef == null) {
+            throw new IndexShardClosedException(shardId);
+        }
+        assert replicationRef.target().getIndexShard().shardId().equals(shardId);
+        return replicationRef;
+    }
+
+//    /** cancel the replication with the given id (if found) and remove it from the replication collection */
```
Comment (Poojita-Raj) [inline server/src/main/java/org/opensearch/indices/replication/copy/ReplicationCollection.java:132]: can replace this commented code with a TODO?

**Review by Poojita-Raj — APPROVED** [review]
LGTM - good work getting it all working!

File: server/src/main/java/org/opensearch/index/engine/EngineConfig.java [hunk 2]
```diff
@@ -95,6 +95,7 @@
     private final CircuitBreakerService circuitBreakerService;
     private final LongSupplier globalCheckpointSupplier;
     private final Supplier<RetentionLeases> retentionLeasesSupplier;
+    private boolean isPrimary;
```
[the remark this answers is not in the document]
Comment (mch2) [reply server/src/main/java/org/opensearch/index/engine/EngineConfig.java:98]: IndexShard actually populates this when wiring up the engine so we can conditionally make InternalEngine readOnly for replicas.  I think a better route here is wiring up a different engine class for replicas.

**Comment by mch2** [comment]
Merging as is and so we can start splitting up issues to iterate on this with #2194

**Pull request merged**